### PR TITLE
add course and subcourse delete along with tests

### PR DESCRIPTION
**Pull request by paul-jaworski**

### CODE REVIEW

**Comment by paul-jaworski** [comment]
I need some help. Where must I delete the subcourse from? Apparantly the subcourse id is referenced in multiple table, i.e.:
grafik

**Comment by Jonasdoubleyou** [comment]
https://github.com/corona-school/backend/blob/master/prisma/schema.prisma every entity that has a courseId or subcourseId

File: graphql/course/mutations.ts [hunk 1]
```diff
@@ -85,6 +85,21 @@ export class MutateCourseResolver {
         return result;
     }
 
+    @Mutation((returns) => GraphQLModel.Course)
+    @AuthorizedDeferred(Role.OWNER)
+    async courseDelete(@Ctx() context: GraphQLContext, @Arg('courseId') courseId: number): Promise<GraphQLModel.Course> {
+        const course = await getCourse(courseId);
+        await hasAccess(context, 'Course', course);
+        if (course) {
```
Comment (Jonasdoubleyou) [inline graphql/course/mutations.ts:93]: How can course be null?

File: graphql/course/mutations.ts [hunk 2]
```diff
@@ -85,6 +85,21 @@ export class MutateCourseResolver {
...
+            const result = await prisma.course.delete({ where: { id: courseId } });
```
Comment (Jonasdoubleyou) [inline graphql/course/mutations.ts:94]: Move this to common/course

Also please implement checks for the prerequisites mentioned in https://github.com/corona-school/project-user/issues/1048

**Review by Jonasdoubleyou — CHANGES_REQUESTED** [review]

**Comment by paul-jaworski** [comment]
@Jonasdoubleyou I need your help :D Somehow the prisma calls to delete subcourses do not get executed. Is there a possibility to execute a single test to observe the sql output?

File: graphql/subcourse/mutations.ts [hunk 3]
```diff
@@ -200,6 +200,19 @@ export class MutateSubcourseResolver {
...
 
+    @Mutation((returns) => Boolean)
+    @AuthorizedDeferred(Role.ADMIN, Role.OWNER)
+    async subcourseDelete(@Ctx() context: GraphQLContext, @Arg('subcourseId') subcourseId: number): Promise<Boolean> {
+        const { user } = context;
+        const subcourse = await getSubcourse(subcourseId);
+        console.log(`Got subcourse ${subcourseId}`);
+        await hasAccess(context, 'Subcourse', subcourse);
+        console.log('We have access');
+        deleteSubcourse(subcourse);
```
Comment (realmayus) [inline graphql/subcourse/mutations.ts:211]: this is missing an `await` @paul-jaworski

File: common/courses/states.ts [hunk 4]
```diff
@@ -88,6 +89,67 @@ export async function denyCourse(course: Course, screeningComment: string | null
...
+    console.log('Deleted lectures');
+    await prisma.subcourse_instructors_student.deleteMany({ where: { subcourseId: subcourse.id } });
+    console.log('Deleted subcourse instructord');
+    await prisma.subcourse_participants_pupil.deleteMany({ where: { subcourseId: subcourse.id } });
+    console.log('Deleted subcourse participants');
+    await prisma.waiting_list_enrollment.deleteMany({ where: { subcourseId: subcourse.id } });
+    console.log('Deleted waiting list enrollment');
+    const deleteResult = await prisma.subcourse.delete({ where: { id: subcourse.id } });
+    console.log(deleteResult);
+    console.log('deleteSubcourse function finished');
```
Comment (realmayus) [inline common/courses/states.ts:150]: Please use `logger.info` for logging

**Review by realmayus — CHANGES_REQUESTED** [review]

**Comment by paul-jaworski** [comment]
So the Delete Course is running through successfully. However it appears that it has sideeffects on other tests (in particular the Create an appointment for a subcourse test.)

Should I remove the Delete Course test or make any other changes or remove it alltogether?

**Comment by Jonasdoubleyou** [comment]
Either reorder the tests so that the course is deleted as the last test, or alternatively create a new course for the test

File: integration-tests/07_course.ts [hunk 5]
```diff
@@ -390,3 +392,40 @@ void test('Add / Remove another instructor', async () => {
         subcourseDeleteInstructor(subcourseId: ${subcourseId} studentId: ${instructor2.student.id})
     }`);
 });
+
+void test('Delete course', async () => {
+    const { courseId, client: courseClient, subcourseId } = await subcourseOne;
+
+    await courseClient
+        .request(
```
Comment (Jonasdoubleyou) [inline integration-tests/07_course.ts:400]: Can just use requestShallFail

File: integration-tests/07_course.ts [hunk 6]
```diff
@@ -6,6 +6,8 @@ import { screenedInstructorOne, screenedInstructorTwo } from './02_screening';
 import { ChatType } from '../common/chat/types';
 import { expectFetch } from './base/mock';
 import { course_coursestate_enum as CourseState } from '@prisma/client';
+import { ValidationError } from 'apollo-server-express';
```
Comment (Jonasdoubleyou) [inline integration-tests/07_course.ts:9]: unused import?

File: graphql/course/mutations.ts [hunk 7]
```diff
@@ -85,6 +85,16 @@ export class MutateCourseResolver {
         return result;
     }
 
+    @Mutation((returns) => Boolean)
+    @AuthorizedDeferred(Role.OWNER)
+    async courseDelete(@Ctx() context: GraphQLContext, @Arg('courseId') courseId: number): Promise<boolean> {
+        const course = await getCourse(courseId);
+        await hasAccess(context, 'Course', course);
+        const result = await deleteCourse(course);
+        logger.info(`Course (${courseId}) deleted successfully`);
```
Comment (Jonasdoubleyou) [inline graphql/course/mutations.ts:94]: I'd move the logging into the function in commons

File: common/courses/states.ts [hunk 8]
```diff
@@ -89,6 +90,51 @@ export async function denyCourse(course: Course, screeningComment: string | null
...
+        throw new ValidationError(`Cannot delete Course ${course.id}, reason: ${can.reason}`);
+    }
+
+    await prisma.course.delete({ where: { id: course.id } });
+}
+
+/* ------------------ Subcourse Delete ------------- */
+
+export function canDeleteSubcourse(subcourse: Subcourse): Promise<Decision> {
+    return new Promise<Decision>((resolve) => {
```
Comment (Jonasdoubleyou) [inline common/courses/states.ts:116]: why the promise?

File: common/courses/states.ts [hunk 9]
```diff
@@ -89,6 +90,51 @@ export async function denyCourse(course: Course, screeningComment: string | null
     logger.info(`Admin denied Course${course.id}) with screening comment: ${screeningComment}`, { courseId: course.id, screeningComment });
 }
 
+/* ------------------ Course Delete ------------- */
+
+export async function canDeleteCourse(course: Course): Promise<Decision> {
+    const subcoursesForCourse = await getSubcoursesForCourse(course.id, false);
+    if (subcoursesForCourse.length == 0) {
+        return { allowed: true, reason: `Course ${course.id} has no subcourses` };
```
Comment (Jonasdoubleyou) [inline common/courses/states.ts:98]: The decision.reason is supposed to be "machine readable", i.e. a constant string where we can then maintain error message translations in the frontend. So it should rather be { allowed: false, reason: "has-subcourse" }, { allowed: true }

**Comment by paul-jaworski** [comment]
> Either reorder the tests so that the course is deleted as the last test, or alternatively create a new course for the test

How can I determinie the execution order of the test? I do not See any test framework being used that I can google for :D

**Comment by Jonasdoubleyou** [comment]
For now test execution is basically sequential, so just create a new file and import it in https://github.com/corona-school/backend/blob/master/integration-tests/index.ts

File: common/courses/states.ts [hunk 10]
```diff
@@ -89,6 +90,50 @@ export async function denyCourse(course: Course, screeningComment: string | null
...
+    const can = await canDeleteSubcourse(subcourse);
+    if (!can.allowed) {
+        throw new ValidationError(`Cannot delete Subcourse ${subcourse.id}, reason: ${can.reason}`);
+    }
+    await prisma.course_participation_certificate.deleteMany({ where: { subcourseId: subcourse.id } });
+    await prisma.lecture.deleteMany({ where: { subcourseId: subcourse.id } });
+    await prisma.subcourse_instructors_student.deleteMany({ where: { subcourseId: subcourse.id } });
+    await prisma.subcourse_participants_pupil.deleteMany({ where: { subcourseId: subcourse.id } });
+    await prisma.waiting_list_enrollment.deleteMany({ where: { subcourseId: subcourse.id } });
+    const deleteResult = await prisma.subcourse.delete({ where: { id: subcourse.id } });
```
Comment (Jonasdoubleyou) [inline common/courses/states.ts:134]: Logging?

File: integration-tests/07_course.ts [hunk 11]
```diff
@@ -390,3 +524,37 @@ void test('Add / Remove another instructor', async () => {
...
+
+    await prisma.subcourse.updateMany({ where: { id: subcourseId }, data: { published: false } });
+
+    await courseClient
+        .request(
+            `
+        mutation DeleteSubcourse {subcourseDelete(subcourseId: ${subcourseId})}      
+    `
+        )
+        .then(async () => {
```
Comment (Jonasdoubleyou) [inline integration-tests/07_course.ts:545]: .then not needed

File: integration-tests/07_course.ts [hunk 12]
```diff
@@ -390,3 +524,37 @@ void test('Add / Remove another instructor', async () => {
...
+
+    //Test deletion of empty course => should work
+
+    await courseClient
+        .request(
+            `
+        mutation DeleteEmptyCourse {courseDelete(courseId: ${courseId})}      
+    `
+        )
+        .then(() => assert.ok('Successfully deleted empty course'))
```
Comment (Jonasdoubleyou) [inline integration-tests/07_course.ts:558]: .then and .catch not needed

**Review by Jonasdoubleyou — COMMENTED** [review]
one last round :)

**Review by Jonasdoubleyou — DISMISSED** [review]

**Review by Jonasdoubleyou — APPROVED** [review]

**Review by Jonasdoubleyou — CHANGES_REQUESTED** [review]
wait, why do you delete .env.example?

**Comment by paul-jaworski** [comment]
> wait, why do you delete .env.example?

ops you are right tnak you

**Review by Jonasdoubleyou — APPROVED** [review]

**Pull request merged**